compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirMemberPropertyChecker.kt

```diff
@@ -5,32 +5,130 @@
 object FirMemberPropertyChecker : FirRegularClassChecker() {
     override fun check(declaration: FirRegularClass, context: CheckerContext, reporter: DiagnosticReporter) {
-        for (member in declaration.declarations) {
-            if (member is FirProperty) {
-                checkProperty(declaration, member, context, reporter)
+        val memberPropertySymbols = declaration.declarations.filterIsInstance<FirProperty>().map { it.symbol }.toSet()
+
+        // A property is known to be initialized only if it is initialized
+        //   1) with its own initializing expression;
+        //   2) at every class constructor;
+        //   3) at any of class's anonymous initializers; or
+        //   4) at other property's initializing expression
+
+        // 2) Property can be initialized at constructors. Since it's unknown what constructor will be used, the property can be determined
+        // as initialized only if it is initialized at every constructor. We should consider a delegated constructor, e.g.,
+        //   constructor() { x = ... }
+        //   constructor(...): this() { ... }  // x will be initialized via this() delegation
+        // We need to topologically sort constructors so that we can process delegated ones before the use sites.
+
+        // 3) Property can be initialized at any of class's anonymous initializers (all of initializers will be executed), e.g.,
+        //   init { x = ... }
+        //   ...
+        //   init { y = ... }
+
+        // 4) Property can be initialized at other property's initializing expression too, e.g.,
+        //   val initX = inlineMe { x = ... } // where inlineMe returns the value of the last expression of the lambda
+
+        // Here, we accumulate initialized info of a property per constructor, anonymous initializer, or property initialization
+        val initializedInConstructor =
+            mutableMapOf<FirPropertySymbol, EventOccurrencesRange>().withDefault { EventOccurrencesRange.ZERO }
+        val initializedInInitOrOtherProperty =
+            mutableMapOf<FirPropertySymbol, EventOccurrencesRange>().withDefault { EventOccurrencesRange.ZERO }
+
+        // To handle the delegated constructor call, we need a cache from constructor to (analyzed) property init info.
+        val constructorToData =
+            mutableMapOf<FirConstructor, PropertyInitializationInfo>().withDefault { PropertyInitializationInfo.EMPTY }
+
+        fun collectInfoFromGraph(
+            graph: ControlFlowGraph,
+            map: MutableMap<FirPropertySymbol, EventOccurrencesRange>,
+            acc: (EventOccurrencesRange, EventOccurrencesRange) -> EventOccurrencesRange,
+            delegatedConstructor: FirConstructor? = null,
+        ) {
+            val delegatedInfo = delegatedConstructor?.let { constructorToData.getValue(it) } ?: PropertyInitializationInfo.EMPTY
+
+            val data = PropertyInitializationInfoCollector(memberPropertySymbols).getData(graph)
```

Comment:
I see that it actually can not be moved to `FirPropertyInitializationAnalyzer`, because this checker requires not only graph of class initializer, but also graphs of constructors too. We discussed this problem internally, and have two solution which may increase performance in this place:
1. Results of `collectInfoFromGraph` can be cached in some service (session component), so actually traversing graph actually will be called only once, and then all checkers may reuse it's results. There are some things to note:
    - we can cache only results for constructors and graph initializers (since functions are not needed here)
    - this service should have special implementation for IDE, because of complex thread model in it (it may be service without any caching at all, actually)
2. Before running CFA we can check properties for some simple invariants (e.g. don't run CFA if all properties of class have initializers). This short pass covers most of real usecases and may dramatically increase performance in this place

Reply:
`FirPropertyInitializationAnalyzer` is about "local" property, while this one is about member/extension properties in classes. The former collects local properties' init info _per_ function, whereas this one collects member properties init info which could be scattered across constructors, anonymous initializers, or other properties' initializing expressions.

I understand constructors and anonymous initializers would be visited twice: one by `FirPropertyInitializationAnalyzer` (for local properties/variables in it), and the other by this checker (for all the member properties). However, due to topological sorting issue, it's quite subtle to _cache_ something here. I don't think we have such mechanism--specifying order of elements to visit---yet, and if we just visit FIR element in the tree without specific order, I think the current approach---collecting info while visiting constructors and anonymous initializers potentially twice---is sound at a minimum. Moreover, I'm not that familiar with thread model in IDE, so I'd like to leave it open for further discussion.

On the other hand, the 2nd bullet point makes sense IIUC. Indeed, if all member properties have their own initializer, we can bail out early, while not collecting/visiting constructors and anonymous initializers, which will improve performance. Added.

Reply:
https://github.com/JetBrains/kotlin/pull/4467 is trying to address the 1st bullet point.